rkt/export.go

```diff
@@ -0,0 +1,237 @@
+// Copyright 2016 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"syscall"
+
+	"github.com/appc/spec/aci"
+	"github.com/appc/spec/schema"
+	"github.com/coreos/rkt/common"
+	"github.com/coreos/rkt/common/overlay"
+	"github.com/coreos/rkt/pkg/user"
+	"github.com/coreos/rkt/store"
+	"github.com/hashicorp/errwrap"
+	"github.com/spf13/cobra"
+)
+
+var (
+	cmdExport = &cobra.Command{
+		Use:   "export UUID OUTPUT_ACI_FILE",
+		Short: "Export an exited pod to an ACI file",
+		Long: `UUID should be the uuid of an exited pod.
+
+Note that currently only pods with a single app can be exported.`,
+		Run: runWrapper(runExport),
+	}
+)
+
+func init() {
+	cmdRkt.AddCommand(cmdExport)
+	cmdExport.Flags().BoolVar(&flagOverwriteACI, "overwrite", false, "overwrite output ACI")
+}
+
+func runExport(cmd *cobra.Command, args []string) (exit int) {
+	if len(args) != 2 {
+		cmd.Usage()
+		return 1
+	}
+
+	aci := args[1]
+	ext := filepath.Ext(aci)
+	if ext != schema.ACIExtension {
+		stderr.Printf("extension must be %s (given %s)", schema.ACIExtension, aci)
+		return 1
+	}
+
+	p, err := getPodFromUUIDString(args[0])
+	if err != nil {
+		stderr.PrintE("problem retrieving pod", err)
+		return 1
+	}
+	defer p.Close()
+
+	if !p.isExited {
+		stderr.Print("pod is not exited. Only exited pods can be exported")
+		return 1
+	}
+
+	apps, err := p.getApps()
+	if err != nil {
+		stderr.PrintE("problem getting the pod's app list", err)
+		return 1
+	}
+
+	if len(apps) != 1 {
+		stderr.Printf("pod has %d apps. Only pods with one app can be exported", len(apps))
+		return 1
+	}
+	app := apps[0]
+
+	if p.usesOverlay() {
+		tmpMntDir := path.Join(getDataDir(), "tmp", fmt.Sprintf("rkt-export-%s", p.uuid))
+		if err := os.MkdirAll(tmpMntDir, common.DefaultRegularDirPerm); err != nil {
+			stderr.PrintE("unable to create temp directory", err)
+			return 1
+		}
+
+		mntDir, err := mountOverlay(p, &app, tmpMntDir)
+		if err != nil {
+			stderr.PrintE(fmt.Sprintf("couldn't mount directory at %s", tmpMntDir), err)
```

Comment:
I think it does not work and it expects the old mounts to still be there:
```
$ sudo rkt export 4185c203 busybox.aci --overwrite
export: error building aci: open /var/lib/rkt/pods/run/4185c203-0005-4c86-84a7-8d4eabbd4c55/stage1/rootfs/opt/stage2/busybox/manifest: no such file or directory
```

To test:
1. start and exit a pod
2. either umount manually the overlayfs or reboot
3. rkt export